bigtop-packages/src/charm/hadoop/layer-hadoop-plugin/reactive/apache_bigtop_plugin.py

```diff
@@ -103,31 +123,44 @@ def send_rm_spec(principal, resourcemanager):
 
 
 @when('apache-bigtop-plugin.yarn.installed')
-@when('hadoop-plugin.joined', 'resourcemanager.ready')
-@when_not('apache-bigtop-plugin.yarn.ready')
-def send_principal_yarn_info(principal, resourcemanager):
+@when('resourcemanager.ready')
+@when('hadoop-plugin.joined')
+def send_plugin_yarn_info(principal, resourcemanager):
     """Send YARN data when the resourcemanager becomes ready."""
     version = get_hadoop_version()
     principal.set_installed(version)
     principal.set_yarn_ready(
         resourcemanager.resourcemanagers(), resourcemanager.port(),
         resourcemanager.hs_http(), resourcemanager.hs_ipc())
-    set_state('apache-bigtop-plugin.yarn.ready')
 
 
-@when('apache-bigtop-plugin.yarn.ready')
+@when_not('resourcemanager.ready')
 @when('hadoop-plugin.joined')
+def clear_plugin_yarn_ready(principal):
+    principal.clear_yarn_ready()
+
+
+@when('resourcemanager.ready')
+@when('hadoop-rest.joined')
+def send_rest_yarn_info(principal, resourcemanager):
+    """Send YARN data when the resourcemanager becomes ready."""
+    principal.set_yarn_ready(
+        resourcemanager.resourcemanagers(), resourcemanager.port(),
+        resourcemanager.hs_http(), resourcemanager.hs_ipc())
+
+
 @when_not('resourcemanager.ready')
-def clear_yarn_ready(principal):
+@when('hadoop-rest.joined')
+def clear_rest_yarn_ready(principal):
     principal.clear_yarn_ready()
-    remove_state('apache-bigtop-plugin.yarn.ready')
-    remove_state('apache-bigtop-plugin.yarn.installed')
 
 
-@when_any('apache-bigtop-plugin.hdfs.installed', 'apache-bigtop-plugin.yarn.installed')
-@when('hadoop-plugin.joined')
```

Comment:
Actually, I think the change to this `update_status` handler's conditions is pointless.  Because it's a subordinate, it will require at minimum one principal, and thus the `hadoop-plugin.joined` state, just for it to exist at all.  Which makes the `hadoop-rest.joined` bit a no-op.  Anyway, it should all work either way.